keps/sig-storage/0035-20190122-inline-csi-volumes-support.md

```diff
@@ -0,0 +1,347 @@
+
+### Specifying allowed inline drivers with `PodSecurityPolicy`
+To control which driver is allowed to be used within a pod spec, this design will update the `PodSecurityPolicy` to introduce `AllowedCSIDrivers` as shown below:
+
+
+```go
+  type PodSecurityPolicySpec struct {
+	// <snip>
+
+	// AllowedCSIDrivers is a whitelist of allowed CSI drivers.  Empty or nil indicates that all
+	// CSI drivers may be used.  This parameter is effective only when the usage of the CSI plugin
+	// is allowed in the "Volumes" field.
+	// +optional
+	AllowedCSIDrivers []AllowedCSIDriver
+  }
+
+  // AllowedCSIDriver represents a single CSI driver that is allowed to be used.
+  type AllowedCSIDriver struct {
+	// Name of the CSI driver
+	Name string
+  }  
+```
+
+Value `PodSecurityPolicy.AllowedCSIDrivers` must be explicitly set with the names of CSI drivers that are allowed to be embedded within a pod spec.  An empty value means no CSI drivers are allowed to be specified inline inside a pod spec.
+
+### Persistent vs Ephemeral lifecycle setting
+A CSI driver must indicate that it supports either persistent or ephemeral inline mode.  This setting will be stored in a [`CSIDriver` configuration CRD](https://github.com/kubernetes/community/pull/2514) and must be retrieved from the API server at runtime.
+
+#### Persistent inline mode
+For CSI drivers that support PV/PVC, the `CSIDriver.lifecyle` value should be set to `persistent` (default).  That value indicates the following:
+
+* The driver uses the normal storage operation flow of *provision*, *attach*,*mount*, *unmount*, *detach*, *delete*.
+* When used in an inline context: 
+  * The `volumeHandle` value must be generated ahead of time and manually entered in the spec 
+  * The driver is only responsible for *attach*, *mount*, *unmount*, *detach*
+  * Provision and deletion are handled outside of the driver
+
+#### Ephemeral inline mode
+Ephemeral volumes are more restricted than their persistent counterparts.  Settiing `CSIDriver.lifecycle = ephemeral` indicates that volumes for the associated CSI driver will follow the lifecycle of its pod.  Ephemeral drivers have the following attributes: 
+* Ephemeral volumes can only work in the context of an inline pod spec
+* The `volumeHandle` is not required and is ignored if provided
+* CSI will internally generate a volumeHandle value which is passed to the driver
+* Ephemeral storage workflow only happens during *mount* and *unmount*
+
+
+### VolumeHandle generation
+When a driver has its lifecycle set to `ephemeral` (see above), the Kubelet (internal CSI code) will employ a naming strategy to generate the value for the volumeHandle.  The generated value will be a combination of `podUID` and `pod namespace`.
```

Comment:
This is not fully unique.

poduid is unique cluster unique. adding namespace doesn't add any uniqueness.

poduid alone is not unique enough though to uniquely identify a unique ephemeral volume inside a pod when multiple are used at once.

poduid + pod.spec.volume[x].name is unique and I recommend we use that.

Consider this example:
```yaml
apiVersion: v1
kind: Pod
metadata:
  name: nginx
spec:
  containers:
  - name: nginx
    image: nginx:1.13-alpine
    ports:
    - containerPort: 80
    volumeMount:
    - name: volume-a
      mountPath: /usr/share/nginx/html/a/
    - name: volume-b
      mountPath: /usr/share/nginx/html/b/
  volumes:
  - name: volume-a
    csi:
      driver: drivers.k8s.io/image
      volumeAttributes:
          image: kfox1111/misc:data-a
  - name: volume-b
    csi:
      driver: drivers.k8s.io/image
      volumeAttributes:
          image: kfox1111/misc:data-b
```

Reply:
> poduid + pod.spec.volume[x].name is unique and I recommend we use that.

Noted.

Reply:
Looks like this still is an issue right here in the doc, in the newest iteration. the rest of the doc looks better though.